<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ MolEvolvR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MolEvolvR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.9</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jravilab/molevolvr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="BinaryDomainNetwork.html">BinaryDomainNetwork()</a></code>

        </dt>
        <dd>Domain Network</dd>
      </dl><dl><dt>

          <code><a href="GCA2Lineage.html">GCA2Lineage()</a></code>

        </dt>
        <dd>Function to map GCA_ID to TaxID, and TaxID to Lineage</dd>
      </dl><dl><dt>

          <code><a href="GenContextNetwork.html">GenContextNetwork()</a></code>

        </dt>
        <dd>Genomic Context Directed Network</dd>
      </dl><dl><dt>

          <code><a href="IPG2Lineage.html">IPG2Lineage()</a></code>

        </dt>
        <dd>IPG2Lineage</dd>
      </dl><dl><dt>

          <code><a href="MolEvolvR_summary.html">summarizeByLineage()</a></code> <code><a href="MolEvolvR_summary.html">summarizeDomArch_ByLineage()</a></code> <code><a href="MolEvolvR_summary.html">summarizeDomArch()</a></code> <code><a href="MolEvolvR_summary.html">summarizeGenContext_ByDomArchLineage()</a></code> <code><a href="MolEvolvR_summary.html">summarizeGenContext_ByLineage()</a></code> <code><a href="MolEvolvR_summary.html">summarizeGenContext()</a></code> <code><a href="MolEvolvR_summary.html">totalGenContextOrDomArchCounts()</a></code>

        </dt>
        <dd>MolEvolvR Summary</dd>
      </dl><dl><dt>

          <code><a href="RepresentativeAccNums.html">RepresentativeAccNums()</a></code>

        </dt>
        <dd>Function to generate a vector of one Accession number per distinct observation from 'reduced' column</dd>
      </dl><dl><dt>

          <code><a href="acc2Lineage.html">acc2Lineage()</a></code>

        </dt>
        <dd>acc2Lineage</dd>
      </dl><dl><dt>

          <code><a href="acc2fa.html">acc2fa()</a></code>

        </dt>
        <dd>acc2fa</dd>
      </dl><dl><dt>

          <code><a href="addLeaves2Alignment.html">addLeaves2Alignment()</a></code>

        </dt>
        <dd>Adding Leaves to an alignment file w/ accessions</dd>
      </dl><dl><dt>

          <code><a href="addLineage.html">addLineage()</a></code>

        </dt>
        <dd>addLineage</dd>
      </dl><dl><dt>

          <code><a href="addName.html">addName()</a></code>

        </dt>
        <dd>Add Name</dd>
      </dl><dl><dt>

          <code><a href="addTaxID.html">addTaxID()</a></code>

        </dt>
        <dd>addTaxID</dd>
      </dl><dl><dt>

          <code><a href="add_leaves.html">add_leaves()</a></code>

        </dt>
        <dd>Adding Leaves to an alignment file w/ accessions</dd>
      </dl><dl><dt>

          <code><a href="add_name.html">add_name()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="advanced_opts2est_walltime.html">advanced_opts2est_walltime()</a></code>

        </dt>
        <dd>Given MolEvolvR advanced options and number of inputs, calculate the total estimated walltime for the job</dd>
      </dl><dl><dt>

          <code><a href="alignFasta.html">alignFasta()</a></code>

        </dt>
        <dd>Perform a Multiple Sequence Alignment on a FASTA file.</dd>
      </dl><dl><dt>

          <code><a href="assign_job_queue.html">assign_job_queue()</a></code>

        </dt>
        <dd>Decision function to assign job queue</dd>
      </dl><dl><dt>

          <code><a href="cleanClusters.html">cleanClusters()</a></code>

        </dt>
        <dd>Cleanup Clust</dd>
      </dl><dl><dt>

          <code><a href="cleanDomainArchitecture.html">cleanDomainArchitecture()</a></code>

        </dt>
        <dd>Cleanup DomArch</dd>
      </dl><dl><dt>

          <code><a href="cleanFAHeaders.html">cleanFAHeaders()</a></code>

        </dt>
        <dd>Cleanup FASTA Header</dd>
      </dl><dl><dt>

          <code><a href="cleanGeneDescription.html">cleanGeneDescription()</a></code>

        </dt>
        <dd>Cleanup GeneDesc</dd>
      </dl><dl><dt>

          <code><a href="cleanGenomicContext.html">cleanGenomicContext()</a></code>

        </dt>
        <dd>Cleanup Genomic Contexts</dd>
      </dl><dl><dt>

          <code><a href="cleanLineage.html">cleanLineage()</a></code>

        </dt>
        <dd>Cleanup Lineage</dd>
      </dl><dl><dt>

          <code><a href="cleanSpecies.html">cleanSpecies()</a></code>

        </dt>
        <dd>Cleanup Species</dd>
      </dl><dl><dt>

          <code><a href="cleanString.html">cleanString()</a></code>

        </dt>
        <dd>Clean String</dd>
      </dl><dl><dt>

          <code><a href="clean_clust_file.html">clean_clust_file()</a></code>

        </dt>
        <dd>Clean Cluster File</dd>
      </dl><dl><dt>

          <code><a href="combine_files.html">combine_files()</a></code>

        </dt>
        <dd>Download the combined assembly summaries of genbank and refseq</dd>
      </dl><dl><dt>

          <code><a href="combine_full.html">combine_full()</a></code>

        </dt>
        <dd>Combining full_analysis files</dd>
      </dl><dl><dt>

          <code><a href="combine_ipr.html">combine_ipr()</a></code>

        </dt>
        <dd>Combining clean ipr files</dd>
      </dl><dl><dt>

          <code><a href="condenseRepeatedDomains.html">condenseRepeatedDomains()</a></code>

        </dt>
        <dd>condenseRepeatedDomains</dd>
      </dl><dl><dt>

          <code><a href="convert2TitleCase.html">convert2TitleCase()</a></code>

        </dt>
        <dd>Changing case to 'Title Case'</dd>
      </dl><dl><dt>

          <code><a href="convertAlignment2FA.html">convertAlignment2FA()</a></code>

        </dt>
        <dd>Adding Leaves to an alignment file w/ accessions</dd>
      </dl><dl><dt>

          <code><a href="convertAlignment2Trees.html">convertAlignment2Trees()</a></code>

        </dt>
        <dd>convertAlignment2Trees</dd>
      </dl><dl><dt>

          <code><a href="convertFA2Tree.html">convertFA2Tree()</a></code>

        </dt>
        <dd>convertFA2Tree</dd>
      </dl><dl><dt>

          <code><a href="convertIPRScanDomainTable2FA.html">convertIPRScanDomainTable2FA()</a></code>

        </dt>
        <dd>Using the table returned from createIPRScanDomainTable, construct a domain fasta for a single accession number in the original fasta (i.e., the original fasta argument to createIPRScanDomainTable())</dd>
      </dl><dl><dt>

          <code><a href="convert_aln2fa.html">convert_aln2fa()</a></code>

        </dt>
        <dd>Adding Leaves to an alignment file w/ accessions</dd>
      </dl><dl><dt>

          <code><a href="countbycolumn.html">countByColumn()</a></code>

        </dt>
        <dd>countByColumn</dd>
      </dl><dl><dt>

          <code><a href="createFA2Tree.html">createFA2Tree()</a></code>

        </dt>
        <dd>createFA2Tree</dd>
      </dl><dl><dt>

          <code><a href="createIPRScanDomainTable.html">createIPRScanDomainTable()</a></code>

        </dt>
        <dd>For a given accession number, get the domain sequences using a interproscan output table &amp; the original FASTA file</dd>
      </dl><dl><dt>

          <code><a href="createJobResultsURL.html">createJobResultsURL()</a></code>

        </dt>
        <dd>Given a pin_id, returns the URL where the user can check the status of their job</dd>
      </dl><dl><dt>

          <code><a href="createJobStatusEmailMessage.html">createJobStatusEmailMessage()</a></code>

        </dt>
        <dd>Produces a mail message that can be sent to a user when their job is accepted. Used by the sendJobStatusEmail() method.</dd>
      </dl><dl><dt>

          <code><a href="createWordCloud2Element.html">createWordCloud2Element()</a></code>

        </dt>
        <dd>Wordclouds for the predominant domains, domain architectures.</dd>
      </dl><dl><dt>

          <code><a href="createWordCloudElement.html">createWordCloudElement()</a></code>

        </dt>
        <dd>Wordclouds for the predominant domains, domain architectures</dd>
      </dl><dl><dt>

          <code><a href="create_lineage_lookup.html">create_lineage_lookup()</a></code>

        </dt>
        <dd>Create a look up table that goes from TaxID, to Lineage</dd>
      </dl><dl><dt>

          <code><a href="domain_network.html">domain_network()</a></code>

        </dt>
        <dd>Domain Network</dd>
      </dl><dl><dt>

          <code><a href="downloadAssemblySummary.html">downloadAssemblySummary()</a></code>

        </dt>
        <dd>Download the combined assembly summaries of genbank and refseq</dd>
      </dl><dl><dt>

          <code><a href="efetchIPG.html">efetchIPG()</a></code>

        </dt>
        <dd>efetchIPG</dd>
      </dl><dl><dt>

          <code><a href="elements2Words.html">elements2Words()</a></code>

        </dt>
        <dd>elements2Words</dd>
      </dl><dl><dt>

          <code><a href="ensureUniqAccNum.html">ensureUniqAccNum()</a></code>

        </dt>
        <dd>make accnums unique</dd>
      </dl><dl><dt>

          <code><a href="extractAccNum.html">extractAccNum()</a></code>

        </dt>
        <dd>extractAccNum</dd>
      </dl><dl><dt>

          <code><a href="filterbydomains.html">filterByDomains()</a></code>

        </dt>
        <dd>filterByDomains</dd>
      </dl><dl><dt>

          <code><a href="filterbyfrequency.html">filterByFrequency()</a></code>

        </dt>
        <dd>filterByFrequency</dd>
      </dl><dl><dt>

          <code><a href="find_top_acc.html">find_top_acc()</a></code>

        </dt>
        <dd>Group by lineage + DA then take top 20</dd>
      </dl><dl><dt>

          <code><a href="findparalogs.html">findParalogs()</a></code>

        </dt>
        <dd>findParalogs</dd>
      </dl><dl><dt>

          <code><a href="formatJobArgumentsHTML.html">formatJobArgumentsHTML()</a></code>

        </dt>
        <dd>Format job arguments into html-formatted key/value pairs, for including in an email</dd>
      </dl><dl><dt>

          <code><a href="gc_undirected_network.html">gc_undirected_network()</a></code>

        </dt>
        <dd>Domain Network</dd>
      </dl><dl><dt>

          <code><a href="generateAllAlignments2FA.html">generateAllAlignments2FA()</a></code>

        </dt>
        <dd>Adding Leaves to an alignment file w/ accessions</dd>
      </dl><dl><dt>

          <code><a href="generate_all_aln2fa.html">generate_all_aln2fa()</a></code>

        </dt>
        <dd>Adding Leaves to an alignment file w/ accessions</dd>
      </dl><dl><dt>

          <code><a href="generate_msa.html">generate_msa()</a></code>

        </dt>
        <dd>Function to generate MSA using kalign</dd>
      </dl><dl><dt>

          <code><a href="getDomainsFromFA.html">getDomainsFromFA()</a></code>

        </dt>
        <dd>getDomainsFromFA</dd>
      </dl><dl><dt>

          <code><a href="getIPRScanColNames.html">getIPRScanColNames()</a></code>

        </dt>
        <dd>Constructor function for interproscan column names (based upon the global variable written in molevol_scripts/R/colnames_molevol.R)</dd>
      </dl><dl><dt>

          <code><a href="getIPRScanColTypes.html">getIPRScanColTypes()</a></code>

        </dt>
        <dd>construct column types for reading interproscan output TSVs (based upon the global variable written in molevol_scripts/R/colnames_molevol.R)</dd>
      </dl><dl><dt>

          <code><a href="get_accnums_from_fasta_file.html">get_accnums_from_fasta_file()</a></code>

        </dt>
        <dd>Get accnums from fasta file</dd>
      </dl><dl><dt>

          <code><a href="get_proc_medians.html">get_proc_medians()</a></code>

        </dt>
        <dd>Scrape MolEvolvR logs and calculate median processes</dd>
      </dl><dl><dt>

          <code><a href="get_proc_weights.html">get_proc_weights()</a></code>

        </dt>
        <dd>Quickly get the runtime weights for MolEvolvR backend processes</dd>
      </dl><dl><dt>

          <code><a href="ipr2viz.html">ipr2viz()</a></code>

        </dt>
        <dd>IPR2Viz</dd>
      </dl><dl><dt>

          <code><a href="ipr2viz_web.html">ipr2viz_web()</a></code>

        </dt>
        <dd>IPR2Viz Web</dd>
      </dl><dl><dt>

          <code><a href="make_opts2procs.html">make_opts2procs()</a></code>

        </dt>
        <dd>Construct list where names (MolEvolvR advanced options) point to processes</dd>
      </dl><dl><dt>

          <code><a href="mapAcc2Name.html">mapAcc2Name()</a></code>

        </dt>
        <dd>Default renameFA() replacement function. Maps an accession number to its name</dd>
      </dl><dl><dt>

          <code><a href="map_acc2name.html">map_acc2name()</a></code>

        </dt>
        <dd>Default rename_fasta() replacement function. Maps an accession number to its name</dd>
      </dl><dl><dt>

          <code><a href="map_advanced_opts2procs.html">map_advanced_opts2procs()</a></code>

        </dt>
        <dd>Use MolEvolvR advanced options to get associated processes</dd>
      </dl><dl><dt>

          <code><a href="msa_pdf.html">msa_pdf()</a></code>

        </dt>
        <dd>Multiple Sequence Alignment</dd>
      </dl><dl><dt>

          <code><a href="plotLineageDA.html">plotLineageDA()</a></code>

        </dt>
        <dd>Lineage Plot: Heatmap of Domains/DAs/GCs vs Lineages</dd>
      </dl><dl><dt>

          <code><a href="plotLineageDomainRepeats.html">plotLineageDomainRepeats()</a></code>

        </dt>
        <dd>Lineage Domain Repeats Plot</dd>
      </dl><dl><dt>

          <code><a href="plotLineageHeatmap.html">plotLineageHeatmap()</a></code>

        </dt>
        <dd>plotLineageHeatmap</dd>
      </dl><dl><dt>

          <code><a href="plotLineageNeighbors.html">plotLineageNeighbors()</a></code>

        </dt>
        <dd>Lineage Plot for top neighbors</dd>
      </dl><dl><dt>

          <code><a href="plotLineageQuery.html">plotLineageQuery()</a></code>

        </dt>
        <dd>Lineage Plot: Heatmap of Queries vs Lineages</dd>
      </dl><dl><dt>

          <code><a href="plotLineageSunburst.html">plotLineageSunburst()</a></code>

        </dt>
        <dd>Lineage Sunburst</dd>
      </dl><dl><dt>

          <code><a href="plotStackedLineage.html">plotStackedLineage()</a></code>

        </dt>
        <dd>Stacked Lineage Plot</dd>
      </dl><dl><dt>

          <code><a href="plotSunburst.html">plotSunburst()</a></code> <code><a href="plotSunburst.html">plotTreemap()</a></code>

        </dt>
        <dd>Create an interactive plotly from count data</dd>
      </dl><dl><dt>

          <code><a href="plotUpSet.html">plotUpSet()</a></code>

        </dt>
        <dd>UpSet Plot</dd>
      </dl><dl><dt>

          <code><a href="plot_estimated_walltimes.html">plot_estimated_walltimes()</a></code>

        </dt>
        <dd>Plot the estimated runtimes for different advanced options and number of inputs</dd>
      </dl><dl><dt>

          <code><a href="prepareColumnParams.html">prepareColumnParams()</a></code>

        </dt>
        <dd>prepareColumnParams</dd>
      </dl><dl><dt>

          <code><a href="prepareSingleColumnParams.html">prepareSingleColumnParams()</a></code>

        </dt>
        <dd>prepareSingleColumnParams</dd>
      </dl><dl><dt>

          <code><a href="proteinAcc2TaxID.html">proteinAcc2TaxID()</a></code>

        </dt>
        <dd>proteinAcc2TaxID</dd>
      </dl><dl><dt>

          <code><a href="proteinAcc2TaxID_old.html">proteinAcc2TaxID_old()</a></code>

        </dt>
        <dd>proteinAcc2TaxID_old</dd>
      </dl><dl><dt>

          <code><a href="readIPRScanTSV.html">readIPRScanTSV()</a></code>

        </dt>
        <dd>Read an interproscan output TSV with standardized column names and types</dd>
      </dl><dl><dt>

          <code><a href="removeAsterisks.html">removeAsterisks()</a></code>

        </dt>
        <dd>Remove Astrk</dd>
      </dl><dl><dt>

          <code><a href="removeEmptyRows.html">removeEmptyRows()</a></code>

        </dt>
        <dd>Remove Empty</dd>
      </dl><dl><dt>

          <code><a href="removeTails.html">removeTails()</a></code>

        </dt>
        <dd>Remove Tails</dd>
      </dl><dl><dt>

          <code><a href="renameFA.html">renameFA()</a></code>

        </dt>
        <dd>Rename the labels of fasta files</dd>
      </dl><dl><dt>

          <code><a href="rename_fasta.html">rename_fasta()</a></code>

        </dt>
        <dd>Rename the labels of fasta files</dd>
      </dl><dl><dt>

          <code><a href="replaceQuestionMarks.html">replaceQuestionMarks()</a></code>

        </dt>
        <dd>Replace QMs</dd>
      </dl><dl><dt>

          <code><a href="reveql.html">reveql()</a></code>

        </dt>
        <dd>reveql</dd>
      </dl><dl><dt>

          <code><a href="reverse_operon.html">reverse_operon()</a></code>

        </dt>
        <dd>reverse_operon</dd>
      </dl><dl><dt>

          <code><a href="runIPRScan.html">runIPRScan()</a></code>

        </dt>
        <dd>runIPRScan</dd>
      </dl><dl><dt>

          <code><a href="run_deltablast.html">run_deltablast()</a></code>

        </dt>
        <dd>Run DELTABLAST to find homologs for proteins of interest</dd>
      </dl><dl><dt>

          <code><a href="run_rpsblast.html">run_rpsblast()</a></code>

        </dt>
        <dd>Run RPSBLAST to generate domain architectures for proteins of interest</dd>
      </dl><dl><dt>

          <code><a href="selectLongestDuplicate.html">selectLongestDuplicate()</a></code>

        </dt>
        <dd>Pick Longer Duplicate</dd>
      </dl><dl><dt>

          <code><a href="sendJobStatusEmail.html">sendJobStatusEmail()</a></code>

        </dt>
        <dd>Sends a "job accepted" email to a user when their job is accepted, including details about the job submission and how to check its status.</dd>
      </dl><dl><dt>

          <code><a href="shortenLineage.html">shortenLineage()</a></code>

        </dt>
        <dd>Shorten Lineage</dd>
      </dl><dl><dt>

          <code><a href="sinkReset.html">sinkReset()</a></code>

        </dt>
        <dd>Sink Reset</dd>
      </dl><dl><dt>

          <code><a href="theme_genes2.html">theme_genes2()</a></code>

        </dt>
        <dd>Theme Genes2</dd>
      </dl><dl><dt>

          <code><a href="to_titlecase.html">to_titlecase()</a></code>

        </dt>
        <dd>To Titlecase</dd>
      </dl><dl><dt>

          <code><a href="validateCountDF.html">validateCountDF()</a></code>

        </dt>
        <dd>validateCountDF</dd>
      </dl><dl><dt>

          <code><a href="wordcloud3.html">wordcloud3()</a></code>

        </dt>
        <dd>plotWordCloud3</dd>
      </dl><dl><dt>

          <code><a href="words2wordcounts.html">words2WordCounts()</a></code>

        </dt>
        <dd>words2WordCounts</dd>
      </dl><dl><dt>

          <code><a href="write.MsaAAMultipleAlignment.html">write.MsaAAMultipleAlignment()</a></code>

        </dt>
        <dd>Write MsaAAMultpleAlignment Objects as algined fasta sequence</dd>
      </dl><dl><dt>

          <code><a href="write_proc_medians_table.html">write_proc_medians_table()</a></code>

        </dt>
        <dd>Write a table of 2 columns: 1) process and 2) median seconds</dd>
      </dl><dl><dt>

          <code><a href="write_proc_medians_yml.html">write_proc_medians_yml()</a></code>

        </dt>
        <dd>Compute median process runtimes, then write a YAML list of the processes and their median runtimes in seconds to the path specified by 'filepath'.</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Janani Ravi, David Mayer, Evan Brenner, Faisal Alquaddoomi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

