# Figures from the MS
## Figure 1. The known Psp systems
The three known Psp operons in E. coli (pspF||pspABC), M. tuberculosis (clgRpspAMN), and B. subtilis (liaIHGFSR) are shown. Key: boxes indicate genes and colors indicate nature of the protein (orange, PspA homolog; blue, transcription factor/regulator; green, transmembrane protein); thin arrows, the direction of transcription; block arrows, promoters. See Introduction.

## Figure 2. The phyletic spread of known Psp members across all major lineages.
a. The known Psp operons in E. coli, M. tuberculosis and B. subtilis resolved into their constituent domain architectures. Key: rectangles, domains; arrow-heads, the direction of transcription; b. Heatmap of the presence of homologs of ‘known’ Psp members. The color gradient represents the number of homologs identified within each lineage (e.g., the darkest shade indicates the highest number of homologs carried by a particular lineage). Columns: Known Psp members — PspA, Snf7, PspB, PspC, PspM, PspN, LiaI, Toast_rack (combining globular domains of LiaG and LiaF) — queried against all sequenced and completed genomes across the three major kingdoms of life. Rows: The major bacterial, archaeal and eukaryotic lineages with representative sequenced genomes (… are not included due to the absence of representative genomes), listed in phyletic order (NCBI taxonomy; ??; ??). The main kingdoms spanning the shown lineages are labeled: bacteria, archaea, and eukaryota. Further details of all homologs and their lineage distributions are shown in Table A1. See Results.

## Figure 3. The genomic contexts containing PspA homologs.
The first row contains domain architectures of the most prevalent homologs of PspA (indicated in light blue). The predominant genomic contexts of PspA homologs across the three major bacterial lineages (Proteobacteria, Firmicutes, and Actinobacteria) have been shown. Representation of the predominant contexts identified by neighborhood searches (±7 genes flanking each homolog). Key: Similar to Fig. 2. Additionally, predominant lineages and the number of times the given context is found in those lineages are indicated in a grey box on the right. Range of numbers given in small black boxes denotes multiple domains co-occurring within a protein/context. Further details of all PspA homologs, their domain architectures, genomic contexts, and lineages are shown in Table 1, A1, A2 and webapp. See Results.

- will add representatives for each DA/GC; GC number to call within the text.

## Figure 4. The domain architectures and genomic contexts underlying all Toast_rack homologs.
a0. Multiple sequence alignment of LiaI-LiaF-TM and Toast_rack domains (B. subtilis). a. Domain architectures of the most prevalent homologs of the Toast_rack domain (indicated in black). Each homolog is a protein comprising the indicated domain architecture (set of domains; each row). b–d. UpSet plot for Toast_rack homologs. b. Distribution (blue histogram) of the constituent domains underlying all Toast_rack homologs c. Variations combinations/configurations in which the key constituent domains come together in the Toast_rack homologs. d. Frequency of occurrences of the different configurations of the Toast_rack homologs (red histogram). e. Predominant genomic contexts of Toast_rack domains. Representation of the predominant contexts identified by neighborhood searches (±7 genes flanking each homolog). Key: Each box indicates a domain and each string of boxes including a directional arrow indicates a protein comprising the indicated domains. The direction of the arrow indicates the direction of transcription based on predicted ORFs (a bidirectional arrow means that the indicated gene could be found in both orientations); X designates absence of indicated genes in some species; ± indicates an optional insertion of one gene. Toast_rack domains are shown in black in each of the configurations. Other domains: Membrane proteins are different shades of orange (SIG, predicted signal peptide, dark orange, PspC , orange, other transmembrane domain, light orange), PspA (blue), known transcription factors/regulators (including HTH, helix-turn-helix domain in teal), DUF, Domain of Unknown Function, and other domains (grey). Further details of all homologs, their domain architectures, genomic contexts, and lineages are shown in Table 1, A1 and webapp. See Results.

- will add i) LiaI-LiaF-TM and Toast_rack domain MSAs;  ii) representatives for each DA/GC & iii) DA/GC numbers for text ref; iv) correct panel #s

## Figure 5. The domain architectures of PspC homologs.
a. Domain architectures of the most prevalent homologs of the PspC (and PspB) domains. b. The phyletic spread of the various PspC domain architectures (see attached note). The color gradient in the heatmap represents the number of homologs identified within each lineage (e.g., the darkest shade indicates the highest number of homologs carried by a particular lineage). Rows: Predominant PspC domain architectures. Columns: Lineages with representative sequenced genomes (… are not included due to the absence of representative genomes), listed in phyletic order (NCBI taxonomy). Further details of all PspC homologs, their domain architectures, genomic contexts, and lineages are shown in Table 1, A1 and webapp. See Results.

- Only DA or DA+GC (those not shown with Toast_rack and PspA)? Show lineage plot with DA or GC? not showing any upset plot. refer to webapp to figure out which is better & what the threshold should be.

## Figure 6. Summary of other key domains found alongside PspA, Toast_rack, and PspC.
- PspMN
- Domain architectures and Genomic contexts. (single lineage)
- DUF1700 and DUF1707; others
- Domain architectures and Genomic contexts — or probably part of the network alone?
- Lineages/phyletic spread, when relevant.
- More details in Table A1 & webapp. See Results.
- temp figures for all the proteins and domains are in the app. will update here finally. some final name cleanup remains!

## Figure 7. Phylogenetic tree of PspA homologs across the tree of life.
a. Representative PspA and Snf7 homologs from all major lineages, domain architectures and genomic contexts (shown in Fig. 3, Table A1, A2) were selected for multiple sequence alignment. b. The phylogenetic tree was constructed based on a multiple sequence alignment performed using PspA homologs across all the major kingdoms/phyla. The key lineages have been labeled next to distinct clusters of similar PspA proteins. The insets are sunburst plots for the homologs of PspA and Snf7 across bacteria, archaea, and eukaryota. The tree leaves are labeled by lineage, species and accession numbers, The outgroup at the bottom of the tree includes Snf7 homologs (indicated in Table A1). See Results.

- add cluster names/numbers; add PDB structures of PspA and Snf7 (here or supplementary?)

## Figure 8. Network & Summary.

- Summarize — multi-panel? PCA kind of summary too? Anything else?
- Companion Table? See Results.
- The figure will be inserted with the final data while MS is being reviewed!



