---
title: "MolEvolvr Help"
subtitle: "_MolEvolvr help docs and tutorials._"
output: html_document
#runtime: shiny
---

```{r setup, include=FALSE, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(fontawesome)
library(tidyverse); library(knitr)
#library(shiny); library(shinydashboard)
```

# MolEvolvR Help docs & Tutorials

## The UI: Workflow and Usage

### How can you use the app?

> what is the end goal?
similar proteins
ancestry of proteins (where they com from, lineage specific, share features with related proteins in different kingdoms)

> exact same da present in other lineages, what could that mean?
  > lineage specific?
  
biology of homologs, what more can they learn from the app than just blast or iprscan alone
if exact same domains are popping up across lineages, what does that say about function?

phyletic spreads and lineages
what are the questions they can answer/analyses they can do?
what is the value?

A critical aspect of understanding the molecular basis of fundamental cellular processes, such as pathogenesis or energy acquisition, is a thorough characterization of the underlying proteins. Several individual computational tools have been developed to characterize proteins at the level of sequence, structure, and function, especially in the context of molecular evolution and phylogeny. However, there still lacks a comprehensive and consolidated workflow for applying all the relevant tools for a protein of interest. We present here the MolEvolvR webapp: a streamlined approach for analyzing proteins in the context of molecular evolution and phylogeny. Our workflow currently starts with one or more proteins of interest -- identifying homologs across the tree of life, followed by a detailed functional characterization in terms of their domain architectures, secondary structure elements, cellular localization -- in the context of phyletic spreads and phylogenetic analyses. This detailed workflow has been implemented in an easy-to-use web application that allows for automated and streamlined molecular evolution and phylogenetic analyses for any protein(s) of interest, including sequence-structure-function relationships as well as lineage-specific adaptations. 

#### Use cases

in broad terms of what we did?
- sample protein(s) -- Psp accessions, homologous and non-homologous
  - tested all analyses on these proteins
  - full analysis on non-homologous proteins
  - domain architecture 


MolEvolvR's functionality is comprised of 4 different types of analysis:

(i) Analysis via Domain Architecture, which allows for exploration of domain interaction and domain propensities;
(ii) Identification of thousands of Homologs; this allows us to see the similarities across various species
(iii) Phylogenetic analysis, which may draw upon information obtained through the other analysis to make inferences and visualizations pertaining to the phyletic spread of the target proteins and their homologs;
(iv) visualizing and analyzing results from other sources, *e.g.,* BLAST suite (BLASTP, BLASTCLUST, RPSBLAST), Interproscan, Multiple Sequence Alignment (MSA).

target audience, why they would use it and how to use the outputs, use cases


## Entering data into the app

### Accession Numbers and Fasta (full analysis)
To begin the full analysis, the accession numbers associated with your proteins of interest can be entered into the "Upload" tab. They should be given in a comma-separated list -- they may be copy/pasted or uploaded as a `.csv` file. You then click on the "Generate FASTA" button and FASTA sequences are retrieved from NCBI [link here]. Additionally, you may give FASTA sequences copy/pasted or as a `.fa` or `.fasta` file.

#### Multiple accessions/FASTA of homologs

`Skip the finding homologs step -- go directly to DA, phylogenetic analysis or only one or the other based on what they choose (checkbox)`

Alternatively, if the user already has a list of homologous proteins, they may enter the workflow post-homology search. Their protein data, in the form of accession numbers or FASTA sequences, is run through our molecular characterization and phylogenetic analysis processes. Additionally, the MolEvolvr webapp can use Clustal Omega, Clustal W, or MUSCLE to perform multiple sequence alignments on the proteins. If the user prefers to use a different alignment algorithm, they may enter their aligned FASTA Sequences that have been aligned using the algorithm of their choosing.

#### Unrelated accessions

`full suite of analysis on each -- each member appears in the dropdown`

A user can start at the beginning of our workflow needing only one or more accession numbers for their protein(s) of interest. Various combinations of the tools can then be selected and the accession numbers will be run through the corresponding procedures in the MolEvol workflow. For instance, the user could perform DeltaBLAST on their input, run the homology search, and then choose to generate the domain architecture using RPSBLAST, InterProScan, or both.

#### Related accessions aka homologs

### Finding homologs

`Queryable interface for homolog data`

Deltablast/psiblast + parameter options To begin the homolog search, a protein's FASTA sequence or accession number is given to the application. If an accession number is given, the app will search for the corresponding FASTA sequence. The FASTA file is then run through either DELTA-BLAST ref or PSI-BLAST ref, which can be chosen by the user. Both are versions of BLASTp ref, with a slight difference to each other: PSI-BLAST runs a BLASTp search and creates a Position-Specific Scoring Matrix (PSSM), which is then used to search the BLAST databases for more matches; DELTA-BLAST also uses PSSMs, but first searches preconstructed PSSMs and the CDD database first before continuing iterations. Once the BLAST homology search completes, it moves on to clustering with BLASTClust ref. Clustering homologs is based on sequence similarity (type of amino acid in the protein) and identity (exactly matching amino acid) to one another, putting proteins that are most similar to one another in a cluster. Cluster names are then added based on the most prominent domain in each group of proteins and the number of proteins in the cluster.

#### Customizing your analysis

------------------------------------------------------------------------

### BLAST outputs from the NCBI BLAST web-interface

Users may also enter the workflow with data obtained from a BLAST run. This data can be run through BLASTCLUST to complete the homology search. The phylogenetic analysis and domain architecture components can then be applied.

The MolEvolvR web-app allows you to input BLAST results that have been run externally on the BLAST web-server for your protein(s) of interest. To help us help you, we've put together a help doc to recommend useful parameters to modify prior to setting up your BLAST runs. Additionally, we'll help you with identifying the right format to download these results in to ensure compatibility with our web-app.

**Here are some instructions on uploading BLASTP results to our "Upload" tab:** <br>

#### STEP 1 Entering accession numbers/fasta sequences and choosing parameters

First, enter your accession number(s) or fasta sequence(s) into the "Enter Query Sequence" box. Choose either the non-redundant (nr) or reference proteins (refseq_proteins) and which algorithm to run. If you would like to further filter your results based on lineages (e.g., species, genus, family, kingdom), please enter the name/taxID accordingly and toggle the box for include/exclude those results in your search. 'blastp' is great for standard protein blast, if you know very little about your protein. If you are interested in identifying remote homologs, we suggest using the PSI-BLAST. DELTA-BLAST works very well if your protein has domains of interest. Creating a job title for the run is optional and for your personal use.

```{r out.width="60%", eval=F}
include_graphics()
```

Choose the number of maximum target sequences to look up. Our recommendation is 5,000--10,000 total target sequences, or hits to ensure maximum inclusion of homologs. Next, go down to the PSI/PHI/DELTABLAST box at the bottom and choose your threshold value, or e-value. We recommend using 1e-5 (also written as 1x10-5 or 0.00001). Double check your parameters in the box at the bottom of the page, then press the BLAST button.

```{r out.width="60%", eval=F}
include_graphics()
```

#### Step 2: Downloading BLAST results

Once the BLAST algorithm has finished running, there will be an option on the upper left of the screen to download results. Click on the "Download" button and select the "Hit Table(text)" option. Once these results are downloaded, you can directly upload these text files to the MolEvolvR web-app.

```{r out.width="60%", eval=F}
include_graphics()
```

```{r out.width="60%", eval=F}
include_graphics()
```

------------------------------------------------------------------------

### InterProScan outputs from the Iprscan5 web-interface

Running InterProScan

#### Step 1 Enter FASTA sequence

Input your protein's fasta sequence by copy/pasting into the box or upload the FASTA file via the "Choose file" button. Double check the sequence is valid with the green check mark next to it. Click the "Search" button to begin the search.

```{r out.width="60%", eval=F}
include_graphics()
```

#### Step 2 Downloading InterProScan results & uploading to MolEvolvR

Download the results in the TSV format under the "Export" dropdown menu. Upload the TSV file to the MolEvolvR app for visualization and further analysis of the domain architectures.

```{r out.width="60%", eval=F}
include_graphics()
```

------------------------------------------------------------------------

## Retrieving results from a prior run
Once submitted
get 8 letter code, go away for 4 hours, come back and enter code, boom results
viz, interactive plots, play around


------------------------------------------------------------------------

/data/research/jravilab/molevolvr_app/scripts/ui/legendText.R <-- use as guide to explain

## Domain Architecture

A protein's domain architecture (DA) refers to the order of specific functional regions of a protein. Two different tools are used by MolEvolvR to characterize the domain architecture of a protein. InterProScan _and RPS-BLAST_ can be chosen by the user to perform the characterization.

`Iprscan + rpsblast, able to choose between the two. Add benchmarking times once that’s fully implemented, add tables/figures`

### Summary stats
A summary table is provided after completion of the analyses. This provides an overview of what domains are conserved across lineages. 

### Phyletic spreads
description

### Upset plots
description of what they are, what does it show/mean

### DA visualizations
description

### Proximity network
description

------------------------------------------------------------------------

## Phylogeny

### Phyletic spread of homologs

### Multiple Sequence Alignments and Phylogenetic Trees

### Paralogs

## Tutorials

------------------------------------------------------------------------

## Contact
Janani Ravi 


