---
title: "JRaviLab"
subtitle: "Software and Data"
output: html_document
#runtime: shiny
editor_options: 
  markdown: 
    wrap: 80
---

```{r setup, include=FALSE, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(fontawesome)
#library(shiny); library(shinydashboard)
```

--------------------------------------------------------------------------------

# [`r fa("users", fill = "#384CB7")`](https://jravilab.github.io)  [`r fa("github", fill = "#1FA67A")`](https://github.com/jravilab) [`r fa("twitter", fill = "#1DA1F2")`](https://twitter.com/jravilab) [`r fa("paper-plane", fill = "red")`](mailto:janani.ravi@cuanschutz.edu)

## *Computational Systems Biology of Infectious Diseases*

We develop discovery platforms that enable experimental and computational
biologists use our approaches to study the molecular basis of pathogenesis and
intervention of infectious diseases.

We welcome collaborations: if you are interested in studying an infectious
disease/pathogen using our platforms or if you would like to see your
computational tool incorporated into our platforms, do [get in
touch](mailto:janani.ravi@cuanschutz.edu).

[**Learn more about our group here.
`r fa("users", fill = "grey20")`**](https://jravilab.github.io)

--------------------------------------------------------------------------------

## Software & Methods

[MolEvolvR](#molevolvr) \| [Connectivity Scores](#connectivity_scores)

### MolEvolvR: A web-app for molecular evolution and phylogeny {#molevolvr}

**MolEvolvR: A web-app for characterizing proteins using molecular evolution and
phylogeny** <br>
*bioRxiv* 2022. doi: <https://doi.org/10.1101/2022.02.18.461833>. <br>
Joseph T Burke\*, Samuel Z Chen\*, Jacob D Krol\*, Lo M Sosinski\*,
Faisal S Alquaddoomi, Vincent P Rubinetti, Evan P Brenner, Ethan P Wolfe,
Shaddai Amolitos, Kellen M Reason, John B Johnson, Janani Ravi^\#^. <br>
[\*Co-primary authors; listed alphabetically. ^\#^Corresponding author:
[janani.ravi\@cuanschutz.edu](mailto:janani.ravi@cuanschutz.edu)] <br>
[`r fa("link", fill = "#1FA67A")`](http://jravilab.org/molevolvr) **Web-app:**
[jravilab.org/molevolvr](http://jravilab.org/molevolvr)

```{r molevolvr, echo=FALSE, eval=F}
shinyAppDir(
  system.file("molevolvr", package = "shiny"),
  options = list(width = "100%", height = 550)
)
```

### Connectivity Scores for Drug Repurposing {#connectivity_scores}

**Reconciling Multiple Connectivity Scores for Drug Repurposing** <br>
Kewalin Samart\*, Phoebe Tuyishime\*,
Arjun Krishnan^\#^, Janani Ravi^\#^. <br>
*Brief. Bioinform.* 2021 Nov 5;22(6):bbab161. doi: <http://doi.org/10.1093/bib/bbab161>. <br>
PMID: 34013329; PMCID: PMC8597919. <br>
[\*Co-primary authors. ^\#^Corresponding authors:
[arjun.krishnan\@cuanschutz.edu](mailto:arjun.krishnan@cuanschutz.edu);
[janani.ravi\@cuanschutz.edu](mailto:janani.ravi@cuanschutz.edu)] <br>
[`r fa("link", fill = "#1FA67A")`](https://jravilab.github.io/connectivity_score_review)
**Live doc:**
[jravilab.github.io/connectivity_score_review](https://jravilab.github.io/connectivity_score_review)

--------------------------------------------------------------------------------

## Data repos & Web-apps

[Bacterial DciA](#dcia_evolution) \| [Listeria Internalins](#inlp_evolution) \|
[Bacterial phage defense](#avcd_evolution) \| [Phage shock stress
response](#psp_evolution)

### Bacterial DciA {#dcia_evolution}

**DciA helicase operators exhibit diversity across bacterial phyla.** <br>
Helen C. Blaine\*, Joseph T. Burke\*, Janani Ravi^\#^, Christina L. Stallings^\#^ <br>
*J. Bacteriology* 2022 Aug 16;204(8):e0016322; doi: <https://doi.org/10.1128/jb.00163-22> <br>
PMID: 35880876; PMCID: PMC9380583. <br>
[\*Co-primary authors. ^\#^Corresponding authors:
[stallings\@wustl.edu](mailto:stallings@wustl.edu);
[janani.ravi\@cuanschutz.edu](mailto:janani.ravi@cuanschutz.edu)] <br>
[`r fa("link", fill = "#1FA67A")`](https://github.com/JRaviLab/dcia_evolution)
**Data:** <https://github.com/JRaviLab/dcia_evolution>

### Internalins in Listeria {#inlp_evolution}

**Novel Internalin P homologs in Listeria ivanovii londoniensis and Listeria seeligeri.** <br>
Kayla N Conner\*, Joseph T Burke\*, Janani Ravi^\#^, Jonathan W Hardy^\#^ <br>
*Microb. Genomics* 2022 Jul;8(7); doi: <https://doi.org/10.1099/mgen.0.000828> <br>
PMID: 35904424; PMCID: PMC9455699. <br>
[\*Co-primary authors. ^\#^Corresponding authors:
[hardyjon\@msu.edu](mailto:hardyjon@msu.edu);
[janani.ravi\@cuanschutz.edu](mailto:janani.ravi@cuanschutz.edu)] <br>
[`r fa("link", fill = "#1FA67A")`](https://github.com/JRaviLab/inlp_listeria)
**Data:** <https://github.com/JRaviLab/inlp_listeria>

### Bacterial phage defense {#avcd_evolution}

**Phage defence by deaminase-mediated depletion of deoxynucleotides in bacteria.** <br>
Hsueh BY\*, Severin GB\*, Elg CA, Waldron EJ, Kant A, Wessel AJ, Dover JA,
Rhoades CR, Ridenhour BJ, Parent KN, Neiditch MB, Ravi J, Top EM, and Waters CM. <br>
*Nat. Microbiol.* 2022 Aug;7(8):1210-1220; doi: <https://doi.org/10.1038/s41564-022-01162-4> <br>
PMID: 35817890; PMCID: PMC9830645. <br>
[\*Co-primary authors. Corresponding author:
[watersc3\@msu.edu](mailto:watersc3@msu.edu)] <br>
[`r fa("link", fill = "#1FA67A")`](https://github.com/JRaviLab/phage_defense_avcd)
**Data:** <https://github.com/JRaviLab/phage_defense_avcd>

### Phage shock protein stress response system {#psp_evolution}

**Phage shock protein (PSP) Envelope Stress Response: Evolutionary History &
Discovery of Novel Players.** <br>
Janani Ravi^\#^, Vivek Anantharaman, Samuel Z Chen, Evan P Brenner,
Pratik Datta, L Aravind^\#^, Maria L Gennaro^\#^. <br>
*bioRxiv* 2020. doi: <https://doi.org/10.1101/2020.09.24.301986> <br>
[^\#^Corresponding authors:
[janani.ravi\@cuanschutz.edu](mailto:janani.ravi@cuanschutz.edu);
[aravind\@nih.gov](mailto:aravind@nih.gov);
[marila.gennaro\@rutgers.edu](mailto:marila.gennaro@rutgers.edu).] <br>
[`r fa("link", fill = "#1FA67A")`](https://jravilab.shinyapps.io/psp-evolution)
**Web-app:**
[jravilab.shinyapps.io/psp-evolution](https://jravilab.shinyapps.io/psp-evolution).
Web-app by JR and SZC.

```{r psp-evolution, echo=FALSE, eval=F}
shinyAppDir(
  system.file("psp-evolution", package = "shiny"),
  options = list(width = "100%", height = 550)
)
```

--------------------------------------------------------------------------------

## Other

**Ten simple rules to host an inclusive conference.** <br>
Rocío Joo\*^\#^, Andrea Sánchez-Tapia\*, Sara
Mortara, Yanina B Saibene, Heather Turner, Dorothea H Peter, Natalia S
Morandeira, Matthias Bannert, Batool Almazrouq, Elizabeth Hare, Laura Ación,
Juan P Narváez-Gómez, Marcela A Córdoba, Federico Marini, Rita Giordano, Silvia
Canelón, Anicet Ebou, Adithi Upadhya, Joselyn Chávez, Janani Ravi^\#^. <br>
*PLoS Comput. Biol.* 2022 Jul;18(7):e1010164.
doi: <https://doi.org/10.1371/journal.pcbi.1010164> <br> 
PMID: 35862309; PMCID: PMC9302732. <br>
[\*Co-primary authors. ^\#^Corresponding authors:
[rocio.joo\@globalfishingwatch.org](mailto:rocio.joo@globalfishingwatch.org);
[janani.ravi\@cuanschutz.edu](mailto:janani.ravi@cuanschutz.edu)] <br>
[`r fa("link", fill = "#1FA67A")`](https://github.com/rociojoo/Ten-simple-rules-inclusive-conference)
**Repo:**
[github.com/rociojoo/Ten-simple-rules-inclusive-conference](https://github.com/rociojoo/Ten-simple-rules-inclusive-conference).
