# Text for the web-app
## Home

## About
- Title, Authors, Affiliations formatted as in the [MS](https://docs.google.com/document/d/1nsduueDAL3QS1E-RBpxJJKGvNmE9371JEaYvkUy5NRg/edit#)
- Abstract
- Links to GitHub + group webpage 
- Maintained by Samuel Zorn Chen, Janani Ravi (w/ my MSU email ID)
- Citation

## Data/Query



## Domain architectures/Genomic contexts
> In each of the following, replace domains with domain architectures & domain architectures w/ genomic contexts for the GC version.

### Lineage ~ Heatmap
The phyletic spread of the various `domain architectures`. The color gradient in the heatmap represents the number of homologs identified within each lineage. Rows: Predominant `domain architectures`. Columns: Key lineages within the three kingdoms of life.

### Summary Table

### UpSet plot
Summary Stats of `Domain Architectures` and underlying `Domains` across all homologs.
Blue Histogram: Distribution of the constituent `domains` underlying `domain architectures`.
Connections: Various combinations/configurations in which the key constituent `domains` come together within a protein homolog.
Red Histogram: Frequency of occurrences of the different domain configurations.

### Wordcloud & Network


### 'As above'

## Other Figs from the Paper
### Fig 1. The phyletic spread of known Psp members across all major lineages.
The three known Psp operons in E. coli (pspF||pspABC), M. tuberculosis (clgRpspAMN), and B. subtilis (liaIHGFSR) are shown. Key: boxes indicate genes and colors indicate nature of the protein (orange, PspA homolog; blue, transcription factor/regulator; green, transmembrane protein); thin arrows, the direction of transcription; block arrows, promoters.

a. The known Psp operons in E. coli, M. tuberculosis and B. subtilis resolved into their constituent domain architectures. Key: rectangles, domains; arrow-heads, the direction of transcription; b. Heatmap of the presence of homologs of ‘known’ Psp members. The color gradient represents the number of homologs identified within each lineage (e.g., the darkest shade indicates the highest number of homologs carried by a particular lineage). Columns: Known Psp members — PspA, Snf7, PspB, PspC, PspM, PspN, LiaI, Toast_rack (combining globular domains of LiaG and LiaF) — queried against all sequenced and completed genomes across the three major kingdoms of life. Rows: The major bacterial, archaeal and eukaryotic lineages with representative sequenced genomes (… are not included due to the absence of representative genomes), listed in phyletic order (NCBI taxonomy; ??; ??). The main kingdoms spanning the shown lineages are labeled: bacteria, archaea, and eukaryota. Further details of all homologs and their lineage distributions are shown in Table A1.

### Fig 2. Phylogenetic tree of PspA homologs across the tree of life.
a. Representative PspA and Snf7 homologs from all major lineages, domain architectures and genomic contexts (shown in Fig. 3, Table A1, A2) were selected for multiple sequence alignment. b. The phylogenetic tree was constructed based on a multiple sequence alignment performed using PspA homologs across all the major kingdoms/phyla. The key lineages have been labeled next to distinct clusters of similar PspA proteins. The insets are sunburst plots for the homologs of PspA and Snf7 across bacteria, archaea, and eukaryota. The tree leaves are labeled by lineage, species and accession numbers, The outgroup at the bottom of the tree includes Snf7 homologs (indicated in Table A1).

- add cluster names/numbers; add PDB structures of PspA and Snf7 (here or supplementary?)

### Fig 3. PspA DA and GC

### Fig 4. w/ `gggenes` or from the MS
Predominant genomic contexts of domains of interest. Representation of the predominant contexts identified by neighborhood searches (±7 genes flanking each homolog). Key: Each box indicates a domain and each string of boxes including a directional arrow indicates a protein comprising the indicated domains. The direction of the arrow indicates the direction of transcription based on predicted ORFs (a bidirectional arrow means that the indicated gene could be found in both orientations); X designates absence of indicated genes in some species; ± indicates an optional insertion of one gene.
_Toast_rack domains are shown in black in each of the configurations. Other domains: Membrane proteins are different shades of orange (SIG, predicted signal peptide, dark orange, PspC , orange, other transmembrane domain, light orange), PspA (blue), known transcription factors/regulators (including HTH, helix-turn-helix domain in teal), DUF, Domain of Unknown Function, and other domains (grey)._

## Phylogeny
### Sunburst

### Tree

### MSA

### Paralog table

