---
title: "The process"
author: "Janani Ravi, Samuel Z Chen"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
    depth: 4
    number_sections: false
    theme: united
---
## Environment & Setup
```{r setup, eval=T, results='hide', message=FALSE}
knitr::opts_knit$set(root.dir=rprojroot::find_rstudio_root_file())
# removed setwd option; homedir of *.Rproj is the correct working dir.
# 
# knitr::purl("docs/the_process.Rmd", "scripts/the_process_20190814.Rmd", documentation = 2)
# infile <- "the_process.Rmd"; inpath <- "docs/"; outpath <- "scripts/"
# name <- strsplit(infile, split = "\\.")[[1]][1]
# knitr::purl(paste0(inpath, infile),
#             paste0(outpath, name,
#                    "_rmd2r_", format(Sys.time(), "%Y%m%d"),
#                    ".R", sep = ""),
#             documentation = 2)
```

```{r load, eval=T, results='hide', message=FALSE}
# library(here)
# library(tidyverse)
# library(rmarkdown); library(knitr)
# library(wordcloud); library(RColorBrewer)
# ADD OTHER DEPENDENCIES

here::here() # checking to see if the homedir location is accurate

# Sourcing scripts containing package functions
source("R/cleanup.R")
# source("R/cleanup_domarch.R")
source("R/clean_clust_file.R")
source("R/summarize.R")
source("R/plotting.R")
source("R/reverse_operons.R")
source("R/add_leaves.R")
source("R/convert_aln2fa.R")
source("R/msa.R")
source("R/tree.R")
# source("R/network.R") # waiting for the new script

# Avoiding conflicts!
conflicted::conflict_prefer("filter", "dplyr")
conflicted::conflict_prefer("as_data_frame", "tibble")
conflicted::conflict_prefer("strsplit", "base")
```

### Cluster files `op_ins_cls` â€”> `tsv`
Run this chunk only if you do not have 'all.txt'
```{r clust_clean_combine, eval=T, results='hide', message=F}

# Initialize dataframe for the combined dataset; Will serve for even 1.
all <- data.frame(matrix(ncol=11, nrow=0))
colnames(all) <- colnames.op_ins_cls

for(x in list.files("data/rawdata_opinscls")){
  print(x)
  inpath <- paste0("data/rawdata_opinscls/", x)
  prot_name <- str_remove(x, ".op_ins_cls")
  # clean_clust_file(path, writepath=NULL, query)
  prot_opinscls <- clean_clust_file(path=inpath, query=prot_name)
  
  all <- bind_rows(all, prot_opinscls)
  # comment next lines to avoid overwriting each time!
  # Last written: Aug 20, 2019
  # write_tsv(prot_opinscls,
  #           path=paste0("data/rawdata_tsv/", prot_name, ".txt"),
  #           col_names=T)
}

# Appending genome-specific signature columns.
# Reading TaxIDs
all_taxid <- read_tsv("data/acc_files/all.taxid", col_names=F)
colnames(all_taxid) <- c("AccNum", "TaxID", "Species.q")
# Reading GCA numbers
all_gca <- read_tsv("data/acc_files/all.gca", col_names=F)
colnames(all_gca) <- c("AccNum", "GCA_ID")
# Adding TaxIDs and GCA_IDs to combined "all" dataframe
all <- all %>%
  left_join(all_taxid, by="AccNum") %>%
  left_join(all_gca, by="AccNum")
# Rearrange columns for easy readability
all <- all %>%
  select(AccNum, Query, ClustID, ClustName,
         DomArch.orig, GenContext.orig,
         Lineage, Species.q, TaxID, GCA_ID,
         arch.PFAM, arch.TMSIG,
         Length, GenName,
         Species.orig, Annotation, GI)

# Write the new combined file; Comment next line(s) to avoid overwriting each time!
# Last written: Aug 20, 2019
# write_tsv(all, path="data/rawdata_tsv/all_raw.txt", col_names=T)
```


## Data import
```{r import, eval=T, results='hide', message=F}
# Example import files: individual or combined ones, pspa.txt, all_raw.txt
prot <- read_delim("data/rawdata_tsv/all_raw.txt",
                   delim="\t", col_names=T, comment="#", trim_ws=T,
                   escape_double=FALSE, na="NA")

lineages_map <- read_delim("data/acc_files/organisms2lineages_map_bae_20170828.txt",
                           delim="\t", col_names=T, comment="#", trim_ws=T)

domains_rename <- read_delim("data/acc_files/domains_rename.txt",
                              delim="\t", col_names=TRUE)

domains_ignore <- read_delim("data/acc_files/domains_ignore.txt",
                             delim="\t", col_names=T)

domains_keep <- read_delim("data/acc_files/domains_keep.txt",
                            delim="\t", col_names=T)

query_domains <- read_delim("data/acc_files/query_domains.txt",
                            delim="\t", col_names=T)
```

## Cleanup
### DomArch cleanup & Reversal of operons
```{r cleanup, eval=T, results='hide', message=F}
# Cleanup Clusters
prot <- prot %>%
  cleanup_clust(TRUE)

# Cleanup Species
prot <- prot %>%
  cleanup_species(FALSE)

# Cleanup GenContext
prot <- prot %>%
  cleanup_gencontext(FALSE)

# Cleanup DomArch
# Rename DomArch.norep to DomArch because that's what replace_doms requires as a column
colnames(prot)[colnames(prot)=="DomArch.norep"] <- "DomArch"

prot <- prot %>%
  cleanup_domarch(TRUE, domains_rename, domains_ignore)

#Give original name back
colnames(prot)[colnames(prot)=="DomArch"] <- "DomArch.norep"

# Reverse/Straighten Operons
prot$GenContext.st <- prot %>%
  reverse_operon()
```

## Data analysis
### Select columns for data export
```{r select-cols, eval=T}
prot_data <- prot %>%
  select(AccNum, Species, Lineage=Lineage.final,
         DomArch=DomArch.norep, GenContext=GenContext.norep,
         Length, GI, GenName, Annotation)
```

### Viewing your data
```{r view-data, eval=T}
paged_table(prot_data)
```

### Wordcounts
```{r wordcounts, eval=T}
## Counts of domains in XXX domain architectures
DA.doms.wc <- prot$DA.doms %>%
  words2wc()
DA.doms.wc %>%
  head() %>%
  kable()
# DA.doms.wc.ge5 <- filter.freq(DA.doms.wc, 5)
## Counts of domain architectures in the GenContext of XXX homologs
GC.DA.wc <- prot$GC.DA %>%
  words2wc()
GC.DA.wc %>%
  head() %>%
  kable()
```

### UpSet plots
```{r upset-plot-da, eval=T, out.height="300px"}
upset.plot(prot, 10, "da2doms")
```

```{r upset-plot-gc, eval=T, out.height="600px"}
upset.plot(prot, 75, "gc2da")
```

### Worclouds
```{r , eval= TRUE, out.height='600px'}
# move to plotting.R and call just the custom function
wordcloud(DA.doms.wc$words, DA.doms.wc$freq, min.freq=1,
          colors=brewer.pal(8, "Spectral"), scale=c(2.5,.5))
wordcloud(GC.DA.wc$words, GC.DA.wc$freq, min.freq=30,
          colors=brewer.pal(8, "Spectral"), scale=c(2.5,.4))
```
### Lineage summaries
```{r lin-summary, eval=T}
## Main Domain Architectures -- Counts by DA and Lineage
prot.DA.summ.byLin <- summ.DA.byLin(prot)
prot.DA.summ <- summ.DA(prot.DA.summ.byLin)

prot.DA.summ.byLin %>% head() %>% kable()
prot.DA.summ %>% head() %>% kable()

## Main Genomic Contexts -- Summarized by DA & Lineage
prot.GC.summ.byDALin <- summ.GC.byDALin(prot)
prot.GC.summ.byLin <- summ.GC.byLin(prot)
prot.GC.summ <- summ.GC(prot.GC.summ.byDALin)

prot.GC.summ.byDALin %>% head() %>% kable()
prot.GC.summ.byLin %>% head() %>% kable()
prot.GC.summ %>% head() %>% kable()
```

### Total Counts
```{r total counts, eval=T, message=FALSE, results='hide'}
DA.cumulative <- total_counts(prot.DA.summ.byLin,
                              cutoff =0, type="DA")
DA.cumulative %>%
  head() %>%
  kable()

GC.cumulative <- total_counts(prot.GC.summ.byDALin,
                              cutoff=50, type="GC") 
GC.cumulative %>%
  head() %>%
  kable()
```
### Lineage summary plots
```{r lin-summ-plot, eval=T, out.width='70%'}
lineage.DA.plot(prot, DA.cumulative,
                "DomArch.norep", "")

lineage.DA.plot(prot, GC.cumulative,
                "GenContext.norep", "")
```

### Add Leaves
```{r add-leaves, eval=T}
add_leaves(input_file ="data/alignments/pspa_snf7.aln",
           lin_file="data/pspa_snf7.txt",
           reduced=TRUE)
```

### Generate MSA + Phylogenetic tree
```{r msa-tree,echo=FALSE,results='hide',fig.keep='all',  out.width="50%",out.height="50%"}
# Create a fasta file
convert_aln2fa(input_file= "data/alignments/pspa_snf7.aln",
               lin_file="data/pspa_snf7.txt",
               output_path="data/alignments/pspa.fasta",
               reduced=TRUE) %>% head()

msa_pdf("data/alignments/pspa.fasta")
pdf_convert(pdf="pspa.fasta.pdf")

files <- list.files(getwd(), pattern ="^pspa.fasta_(.*).png$")
include_graphics(files)

#pdf2image then render image here?
#should I be deleting images
seq_tree("data/alignments/pspa.fasta")
```




