---
title: "The process"
author: "Janani Ravi"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
    depth: 4
    number_sections: false
    theme: united
---
## Environment & Setup
```{r setup, eval=T, results='hide', message=FALSE}
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file(), setwd(".."))

library(tidyverse)
library(here)
library(rmarkdown); library(knitr)
library(wordcloud); library(RColorBrewer)
#####TEMP#####
#setwd("..")

source("R/cleanup.R")
source("R/summarize.R")
source("R/plotting.R")
source("R/reverse_operons.R")
source("R/add_leaves.R")
source("R/convert_aln2fa.R")
source("R/msa.R") 
source("R/tree.R");
conflicted::conflict_prefer("filter", "dplyr")
conflicted::conflict_prefer("as_data_frame", "tibble")
conflicted::conflict_prefer("strsplit","base")
```


## Data import
```{r import, eval=T, results='hide', message=F}
prot <- read_delim("data/pspa_snf7.txt", # data/protein_name.txt
                   delim="\t", col_names=T, comment="#", trim_ws=T,
                   escape_double = FALSE, na = "NA")

lineages_map <- read_delim("data/organisms2lineages_map_bae_20170828.txt",
                           delim="\t", col_names=T, comment="#", trim_ws=T)

domains.replace <- read_delim("data/domains.replace.txt",
                                delim = "\t", col_names = TRUE)

domains.remove <- read_delim("data/domains.ignore.txt",
                             delim = "\t", col_names = TRUE)
```


## Cleanup
```{r import, eval=T, results='hide', message=F}
prot <- prot %>%
  reverse_operons() %>%
  cleanup_species()
#Rename DomArch.norep to DomArch because that's what replace_doms requires as a column
colnames(prot)[colnames(prot)=="DomArch.norep"] <- "DomArch"
  replace_doms(prot,domains.replace,domains.remove)
#Give original name back
colnames(prot)[colnames(prot)=="DomArch"] <- "DomArch.norep"
```

## Data analysis
### Select columns for data export
```{r select-cols, eval=T}
prot_data <- prot %>%
  select(AccNum, Species, Lineage=Lineage.final,
         DomArch = DomArch.norep, GenContext=GenContext.norep,
         Length, GI, GenName, Annotation)
```

### Viewing your data
```{r view-data, eval=T}
paged_table(prot_data)
```

### Wordcounts
```{r wordcounts, eval=T}
## Counts of domains in XXX domain architectures
DA.doms.wc <- prot$DA.doms %>%
  words2wc()
DA.doms.wc %>%
  head() %>%
  kable()
# DA.doms.wc.ge5 <- filter.freq(DA.doms.wc, 5)
## Counts of domain architectures in the GenContext of XXX homologs
GC.DA.wc <- prot$GC.DA %>%
  words2wc()
GC.DA.wc %>%
  head() %>%
  kable()
```



### UpSet plots
```{r upset-plot-da, eval=T, out.height="300px"}
upset.plot(prot, 10, "da2doms")
```

```{r upset-plot-gc, eval=T, out.height="600px"}
upset.plot(prot, 75, "gc2da")
```

### Worclouds
```{r , eval= TRUE, out.height='600px'}
wordcloud(DA.doms.wc$words,DA.doms.wc$freq,min.freq = 1,colors=brewer.pal(8, "Spectral"),scale=c(2.5,.5))
wordcloud(GC.DA.wc$words,GC.DA.wc$freq,min.freq = 30,colors=brewer.pal(8, "Spectral"), scale=c(2.5,.4))
```
### Lineage summaries
```{r lin-summary, eval=T}
## Main Domain Architectures -- Counts by DA and Lineage
prot.DA.summ.byLin <- summ.DA.byLin(prot)
prot.DA.summ <- summ.DA(prot.DA.summ.byLin)

prot.DA.summ.byLin %>% head() %>% kable()
prot.DA.summ %>% head() %>% kable()

## Main Genomic Contexts -- Summarized by DA & Lineage
prot.GC.summ.byDALin <- summ.GC.byDALin(prot)
prot.GC.summ.byLin <- summ.GC.byLin(prot)
prot.GC.summ <- summ.GC(prot.GC.summ.byDALin)

prot.GC.summ.byDALin %>% head() %>% kable()
prot.GC.summ.byLin %>% head() %>% kable()
prot.GC.summ %>% head() %>% kable()
```

### Total Counts
```{r total counts, eval=T, message=FALSE, results='hide'}
DA.cummulative <- total_counts(prot.DA.summ.byLin, cutoff =0, type = "DA")
DA.cummulative %>%
  head() %>%
  kable()

GC.cummulative <- total_counts(prot.GC.summ.byDALin, cutoff = 50, type = "GC") 
GC.cummulative %>%
  head() %>%
  kable()




```
### Lineage summary plots
```{r lin-summ-plot, eval=T, out.width='70%'}
lineage.DA.plot(prot, DA.cummulative,"DomArch.norep", "")

lineage.DA.plot(prot, GC.cummulative,"GenContext.norep", "")
```

### Add Leaves
```{r add-leaves, eval=T}
add_leaves(input_file ="data/alignments/pspa_snf7.aln",lin_file="data/pspa_snf7.txt",reduced = TRUE)
```

### Generate MSA + Phylogenetic tree
```{r msa-tree,echo=FALSE,results='hide',fig.keep='all',  out.width="50%",out.height="50%"}
#Create a fasta-file
convert_aln2fa(input_file= "data/alignments/pspa_snf7.aln",lin_file = "data/pspa_snf7.txt",output_path = "data/alignments/pspa.fasta", reduced = TRUE) %>% head()

msa_pdf("data/alignments/pspa.fasta")
pdf_convert(pdf = "pspa.fasta.pdf")

files <- list.files(getwd(), pattern ="^pspa.fasta_(.*).png$")
include_graphics(files)



#pdf2image then render image here?
#should I be deleting images
seq_tree("data/alignments/pspa.fasta")
```


  
  